File: java/day19_part1_2017.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Main {
    public static void main(String[] args) throws IOException {
        try (BufferedReader br = new BufferedReader(new FileReader("input.txt"))) {
            String line;
            StringBuilder sb = new StringBuilder();
            int x = 0, y = 0;
            String[] grid = new String[1000]; // assuming max 1000 lines
            int i = 0;
            while ((line = br.readLine()) != null) {
                grid[i++] = line;
                if (x == 0) {
                    for (int j = 0; j < line.length(); j++) {
                        if (line.charAt(j) == '|') {
                            x = j;
                            break;
                        }
                    }
                }
            }

            int dx = 0, dy = 1;
            StringBuilder letters = new StringBuilder();

            while (true) {
                if (x < 0 || x >= grid[0].length() || y < 0 || y >= grid.length) {
                    break;
                }

                char cell = grid[y].charAt(x);

                if (cell == ' ') {
                    break;
                }

                if (cell >= 'A' && cell <= 'Z') {
                    letters.append(cell);
                }

                if (cell == '+') {
                    if (dx == 0) {
                        if (x > 0 && (grid[y].charAt(x - 1) == '-' || (grid[y].charAt(x - 1) >= 'A' && grid[y].charAt(x - 1) <= 'Z'))) {
                            dx = -1;
                            dy = 0;
                        } else {
                            dx = 1;
                            dy = 0;
                        }
                    } else {
                        if (y > 0 && (grid[y - 1].charAt(x) == '|' || (grid[y - 1].charAt(x) >= 'A' && grid[y - 1].charAt(x) <= 'Z'))) {
                            dx = 0;
                            dy = -1;
                        } else {
                            dx = 0;
                            dy = 1;
                        }
                    }
                }

                x += dx;
                y += dy;
            }

            System.out.println(letters.toString());
        }
    }
}

File: matlab/day19_part1_2017.m

function solve()
    fid = fopen('input.txt', 'r');
    if fid == -1
        error('Could not open input.txt');
    end

    lines = {};
    maxLen = 0;
    tline = fgetl(fid);
    while ischar(tline)
        lines{end+1} = tline;
        maxLen = max(maxLen, length(tline));
        tline = fgetl(fid);
    end
    fclose(fid);

    rows = length(lines);
    cols = maxLen;

    grid = repmat(' ', rows, cols);
    for i = 1:rows
        line = lines{i};
        grid(i, 1:length(line)) = line;
    end

    y = 1;
    x = find(grid(y, :) == '|', 1);

    dx = 0;
    dy = 1;

    letters = '';

    while true
        if y < 1 || y > rows || x < 1 || x > cols
            break;
        end

        cellChar = grid(y, x);

        if cellChar == ' '
            break;
        end

        if cellChar >= 'A' && cellChar <= 'Z'
            letters(end+1) = cellChar;
        end

        if cellChar == '+'
            if dx == 0
                isLeftPath = (x > 1 && (grid(y, x-1) == '-' || (grid(y, x-1) >= 'A' && grid(y, x-1) <= 'Z')));
                if isLeftPath
                    dx = -1;
                else
                    dx = 1;
                end
                dy = 0;
            else
                isUpPath = (y > 1 && (grid(y-1, x) == '|' || (grid(y-1, x) >= 'A' && grid(y-1, x) <= 'Z')));
                if isUpPath
                    dy = -1;
                else
                    dy = 1;
                end
                dx = 0;
            end
        end

        x = x + dx;
        y = y + dy;
    end

    disp(letters);
end
